prometheus/filter.go

```diff
@@ -0,0 +1,117 @@
+package prometheus
+
+import (
+	"sort"
+
+	proto "github.com/golang/protobuf/proto"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+)
+
+var _ prometheus.Gatherer = (*Filter)(nil)
+
+// Filter filters the metrics from Gather using Matcher.
+type Filter struct {
+	Gatherer prometheus.Gatherer
+	Matcher  Matcher
+}
+
+// Gather filters all metrics to only those that match the Matcher.
+func (f *Filter) Gather() ([]*dto.MetricFamily, error) {
+	mfs, err := f.Gatherer.Gather()
+	if err != nil {
+		return nil, err
+	}
+	return f.Matcher.Match(mfs), nil
+}
+
+// Matcher is used to match families of prometheus metrics.
+type Matcher map[string]Labels // family name to label/value
+
+// NewMatcher returns a new matcher.
+func NewMatcher() Matcher {
+	return Matcher{}
+}
+
+// Family helps constuct match by adding a metric family to match to.
+func (m Matcher) Family(name string, lps ...*dto.LabelPair) Matcher {
```

Comment:
This returns `m` as it's expected that people will do chaining off of `Family` correct? like

```go
m.Family("foo",lps..).Family("bar",...)
```

Reply:
Yup exactly ... example is now in promtheus/example_test.go